README.md
# OpenAPI Factory (Javascript)

[![npm version](https://badge.fury.io/js/openapi-factory.svg)](https://badge.fury.io/js/openapi-factory)

Provides an simple event interceptor for AWS Lambda. Handles events tha come from API Gateway, SQS, EventBridge, and all other AWS services and routes them to appropriate lambda code. This enables you to have single AWS Lambda function for your whole service instead of needing tons of nano-ones to handle every route.

By merging your lambda functions together, you eliminate 99% of all cold starts, and simplify deployment to AWS Lambda.

## Partner libraries
Each of these usable completely independently. But both work together as well:

With the (aws-architect)[https://www.npmjs.com/package/aws-architect] npm package, you can also develop your lambda locally, spinning up a full HTTP API to have quick development cycles.
* The **OpenAPI Factory** provides the production runtime wrapper to convert all the different AWS events into a simple format and processes the result
* The **AWS Architect** library, let's you build, test, and run locally your lambda. And when you are ready it automatically packages your lambda and publishes it in S3, making it ready to pull into your Infrastructure as Code solution as soon as you would like.

### Create an API

```js
const ApiFactory = require('openapi-factory');
const options = {
  requestMiddleware(request, context) {
    return request;
  },
  responseMiddleware(request, response) {
    return response;
  },
  errorMiddleware(request, error) {
    return { statusCode: 500, body: { message: 'Unexpected Error' } };
  }
};
const api = new ApiFactory(options);

api.get('/example', async request => {
  // Short hand for returning a JSON object
  return { value: 'test' };

  // Or explicitly return the whole response
  return {
    body: { value: 'testWithStatus' },
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' }
  };
});

// converts dynamic variables paths
api.get('/example/{id}/subpath', request => {
  const idFromPath = request.pathParameters.id;
  const stageVariable = request.stageVariables.VARIABLE_NAME;
  const query = request.queryStringParameters.QUERY_NAME;
  const headers = request.headers.HEADER_NAME;
});

api.setAuthorizer(request => {
  return 'valid-policy-document';
});

api.onEvent(event => {
  console.log('triggered by a direct invocation from a Lambda Event Source.');
// AWS Documentation: https://docs.aws.amazon.com/lambda/latest/dg/lambda-services.html
// Example payloads: https://lambda.101i.de/
});

api.onSchedule(data => {
  console.log('triggered by a CloudWatch Rule schedule');
});

api.get('/items/{itemid}', async request => {
  console.log(request.path.itemId);
  return {
    body: { value: 'testWithStatus' },
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' }
  };
});

// paths have an optional options object which has property "rawBody" to return the raw body only.
api.get('/items/{itemid}', { rawbody: true }, async request => {
  console.log('This is he raw body of the request: ', request.body);
  return { statusCode: 200 };
});

// Example: AWS Api Gateway magic string handling for CORS and 404 fallbacks.
api.options('/{proxy+}', () => {
  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key',
      'Access-Control-Allow-Methods': 'DELETE,GET,HEAD,OPTIONS,PATCH,POST,PUT',
      'Access-Control-Allow-Origin': '*'
    }
  };
});

api.any('/{proxy+}', () => {
  return {
    statusCode: 404,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*'
    }
  };
});

```

### Default Headers
The default headers returned unless overwritten are:
* For a JSON Object: `{ 'Content-Type': 'application/links+json', 'Access-Control-Allow-Origin': '*' }`
* For a binary Object: `{ 'Content-Type': 'application/octet-stream', 'Access-Control-Allow-Origin': '*' }`

#### Custom PathResolver
It is possible that the default handling of REST routes does not match explicitly match your strategy for resolution. Since there is nothing more than string matching it is fairly easy to hoist this function.

```js
const ApiFactory = require('openapi-factory');
// The default path resolver is the one contained in the PathResolver.
// * It parses the registered paths, stores them in a dictionary, and then looks them up later when necessary.
options.pathResolver = new PathResolver();
const api = new ApiFactory(options);

// However this can be replaced by a custom implementation which includes storePath and resolvePath
class PathResolver {
  constructor() {
    this.currentPathDictionary = {
      GET: {},
      POST: {}
      // ...
    };
  }

  // will get in the current dictionary object as well, there is a specific dictionary for each verb
  // * the current path string
  // * the object associated with that path
  // and returns the updated dictionary
  storePath(currentVerbPathDictionary, dynamicPath, storageObject) {
    return new PathDictionary();
  }

  // Later resolvePath is called to get back the currentPathDictionary and raw path,
  // * and expect to return the pre-stored storageObject
  resolvePath(currentPathDictionary, resolvedPath) {
    return storageObject;
  }
}
```

## Lambda@Edge example
[See Example here](./lambda@edge-cloudfront-wrapper.md)


CHANGELOG.md
# Change log
This is the changelog for [OpenAPI Factory](readme.md).

## 5.4 ##
* Invert the path map. This actually isn't a breaking change, but you are directly using the internals of this library, they are changing.
* 204 & 304 response status codes shall never include the body.

## 5.3 ##
* Upgrade to node 14

## 5.2 ##
* Enable returning strings as response bodies.
* Support API Gateway HTTP API (v2).
* Support `QUERY` method verb.

## 5.1 ##
* Always trust APIGW `event.path` value and don't try to overwrite it.

## 5.0 ##
* Upgrade to node 12
* Add greedy matching by add `+` at the end of any dynamic resource replacement.
* Pass api route options back to the event for reuse in middlewares

## 4.4 ##
* Provide a way to override the DefaultPathResolver
* Add `request.route` to be the dynamic path
* Provide `context` as part of the requestMiddleware
* Automatically url decode path parameters.
* Allow `*` to be a valid path parameter
* Populate multiValueHeaders instead of headers to support header lists

## 4.3 ##
* Return `null` bodies when body is not specified rather than defaulting `{}`
* Provide the `rawBody` options parameter to keep the original body with the request instead of converting to JSON.

## 4.2 ##
* Include middleware for requests and responses
* Always fallback to event handling when neither API or schedule is triggered.
* When a path token is not specified the value of that token is `null`.

## 4.1 ##
* Optimize code to use es6 classes

## 4.0 ##
* Now comes with the `async` handler, feel free to await all the calls.
* Default
```json
    {
        "Content-Type": "application/link+json",
        "Access-Control-Allow-Origin": "*"
    }
```
## 3.2 ##
* Prevent registration of duplicate paths

## 3.1 ##
* Provide defaults for event parameters in the path, query, stage, and headers well not specified.
* Added two new triggers `onEvent` and `onSchedule` which will get triggered via the appropriate mechanisms.

## 3.0 ##
* Change AWS authorizer type to REQUEST from TOKEN.
* Improve logging for request failures.

## 2.2 ##
* paths can be specified at the top level instead of needing to put the paths in '{/proxy+}'.

## 2.1 ##
* Binary data can now be sent as a buffer.

## 2.0 ##
* Support new API Gateway format.
* Set default AuthorizerPolicy with promises.

## 1.1 ##
* Convert to using singlu function to Lambda Functions will work with the API without class instantiation.
* Add VPC Security Group and Subnet options.
* Remove reference to configuration (moved to deployment configuration instead of API configuration).

## 1.0 ##
* Created npm library for building apis.
* Pull api method and resource from AWS context.
* Standardize names of ResourcePath, and Method


tests/make.js
'use strict;'
var esprima = require('esprima');
var mocha = require('mocha');
var assert = require('chai').assert;
var fs = require('fs');
var path = require('path');

describe('make.js', function() {
	describe('Syntax', function () {
		it('Should be valid Javascript', function() {
			try {
				var userStringToTest = fs.readFileSync(path.resolve('make.js'));
				esprima.parse(userStringToTest);
				assert(true);
			}
			catch(e) {
				console.error(e);
				assert(false, JSON.stringify(e, null, 2));
			}
		});
		it('Should be valid node', function(){
			try {
				var app = require('../make');
				assert(true);
			}
			catch(e) {
				console.error(e);
				assert(false, JSON.stringify(e, null, 2));
			}
		});
	});
});

tests/index.js
'use strict;'
var esprima = require('esprima');
var mocha = require('mocha');
var assert = require('chai').assert;
var fs = require('fs');
var path = require('path');

describe('index.js', function() {
	describe('Syntax', function() {
		it('Should be valid Javascript', function() {
			try {
				var userStringToTest = fs.readFileSync(path.resolve('index.js'));
				esprima.parse(userStringToTest);
				assert(true);
			}
			catch(e) {
				console.error(e);
				assert(false, JSON.stringify(e, null, 2));
			}
		});
		it('Should be valid node', function() {
			try {
				var app = require('../index');
				assert(true);
			}
			catch(e) {
				console.error(e);
				assert(false, JSON.stringify(e, null, 2));
			}
		});
	});
	describe('methods', function() {
		it('Check expected API Methods', function() {
			var isFunction = (obj) => { return !!(obj && obj.constructor && obj.call && obj.apply); };
			var Api = require('../index');
			var api = new Api();
			assert.isFunction(api.head, 'HEAD has not been defined.');
			assert.isFunction(api.get, 'GET has not been defined.');
			assert.isFunction(api.put, 'PUT has not been defined.');
			assert.isFunction(api.patch, 'PATCH has not been defined.');
			assert.isFunction(api.post, 'POST has not been defined.');
			assert.isFunction(api.delete, 'DELETE has not been defined.');
			assert.isFunction(api.any, 'ANY has not been defined.');
		});
	});
	describe('authorizer', function() {
		it('check call to default authorizerFunc', function() {
			try {
				var Api = require('../index');
				var api = new Api();
				assert(api.Authorizer.AuthorizerFunc === null);
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
		it('check call to false authorizerFunc', function() {
			try {
				var Api = require('../index');
				var api = new Api();
				api.SetAuthorizer(() => false);
				var result = api.Authorizer.AuthorizerFunc();
				//result should be false;
				assert(!result);
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
		it('check call to false authorizerFunc', function(done) {
			try {
				var Api = require('../index');
				var api = new Api();
				api.SetAuthorizer(() => {
					return Promise.reject('Unauthorized');
				});
				var result = api.Authorizer.AuthorizerFunc()
				.then(bad => {
					done('Should have failed');
				}, correctFailure => {
					done();
				});
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
		it('check call to success authorizerFunc', function(done) {
			try {
				var Api = require('../index');
				var api = new Api();
				api.SetAuthorizer(() => {
					return Promise.resolve();
				});
				var result = api.Authorizer.AuthorizerFunc()
				.then(success => {
					done();
				}, correctFailure => {
					done(`Should not have failed: ${correctFailure}`);
				});
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
		it('check call to success authorizer', function(done) {
			try {
				var Api = require('../index');
				var api = new Api();
				api.SetAuthorizer(() => {
					return Promise.resolve();
				});
				var result = api.Authorizer.AuthorizerFunc()
				.then(success => {
					done();
				}, incorrectFailure => {
					done(`Should not have failed: ${incorrectFailure}`);
				});
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
		it('check call to failure authorizer', function(done) {
			try {
				var Api = require('../index');
				var api = new Api();
				api.SetAuthorizer(() => {
					return Promise.reject('Fail this test');
				});
				api.Authorizer.AuthorizerFunc()
				.then(incorrectSuccess => {
					done(`Should not have passed: ${incorrectSuccess}`);
				}, correctFailure => {
					done();
				});
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
		it('check call to failure authorizer handler', function(done) {
			try {
				var Api = require('../index');
				var api = new Api();
				api.SetAuthorizer(() => {
					return Promise.reject('Fail this test');
				});
				api.handler({
					type: 'TOKEN',
					authorizationToken: 'token',
					methodArn: 'authorizationHandlerTest'
				}, {}, (successfulFailure, incorrectSuccess) => { incorrectSuccess ? done('This test should have failed') : done(); });
			}
			catch(e) {
				console.error(e);
				assert(false, e.toString());
			}
		});
	});
	describe('handler', function() {
		it('check call to ANY handler', function(done) {
			try {
				var expectedResult = {'Value': 5};
				var Api = require('../index');
				var api = new Api();
				api.any('/test', (request) => {
					return new Api.Response(expectedResult);
				});

				api.handler({
					httpMethod: 'GET',
					resource: '/test'
				}, {}, (_, x) => x)
				.then(output => {
					assert.deepEqual(JSON.parse(output.body), expectedResult, `Output data does not match expected.`);
					assert.strictEqual(output.statusCode, 200, 'Status code should be 200');
					done();
				})
				.catch(failure => done(failure));
			}
			catch(e) {
				console.error(e.stack);
				assert(false, e.toString());
			}
		});
		it('check call to GET handler', function(done) {
			try {
				var expectedResult = {'Value': 5};
				var Api = require('../index');
				var api = new Api();
				api.get('/test', (request) => {
					return new Api.Response(expectedResult);
				});

				api.handler({
					httpMethod: 'GET',
					resource: '/test'
				}, {}, (_, x) => x)
				.then(output => {
					assert.deepEqual(JSON.parse(output.body), expectedResult, `Output data does not match expected.`);
					assert.strictEqual(output.statusCode, 200, 'Status code should be 200');
					done();
				})
				.catch(failure => done(failure));
			}
			catch(e) {
				console.error(e.stack);
				assert(false, e.toString());
			}
		});
		it('check promise result to GET handler', function(done) {
			try {
				var expectedResult = {'Value': 5};
				var Api = require('../index');
				var api = new Api();
				api.get('/test', (request) => {
					return Promise.resolve(new Api.Response(expectedResult));
				});

				api.handler({
					httpMethod: 'GET',
					resource: '/test'
				}, {}, (_, x) => x)
				.then(output => {
					assert.deepEqual(JSON.parse(output.body), expectedResult, `Output data does not match expected.`);
					assert.strictEqual(output.statusCode, 200, 'Status code should be 200');
					done();
				})
				.catch(failure => done(failure));
			}
			catch(e) {
				console.error(e.stack);
				assert(false, e.toString());
			}
		});
		it('check promise rejection to GET handler', function(done) {
			try {
				var expectedResult = {'Value': 5};
				var Api = require('../index');
				var api = new Api();
				api.get('/test', (request) => {
					return Promise.reject(expectedResult);
				});

				api.handler({
					httpMethod: 'GET',
					resource: '/test'
				}, {}, (_, x) => x)
				.then(output => {
					assert.deepEqual(JSON.parse(output.body), expectedResult, `Output data does not match expected.`);
					assert.strictEqual(output.statusCode, 500, 'Promise rejections should be 500');
					done();
				})
				.catch(failure => done(failure));
			}
			catch(e) {
				console.error(e.stack);
				assert(false, e.toString());
			}
		});
		it('check exception in GET handler', function(done) {
			try {
				var expectedResult = {'Value': 5};
				var Api = require('../index');
				var api = new Api();
				api.get('/test', (request) => {
					throw expectedResult;
				});

				api.handler({
					httpMethod: 'GET',
					resource: '/test'
				}, {}, (_, x) => x)
				.then(output => {
					assert.deepEqual(JSON.parse(output.body), expectedResult, `Output data does not match expected.`);
					assert.strictEqual(output.statusCode, 500, 'Error should be a 500 on a throw');
					done();
				})
				.catch(failure => done(failure));
			}
			catch(e) {
				console.error(e.stack);
				assert(false, e.toString());
			}
		});
		it('check Error in GET handler', function(done) {
			try {
				var errorMessage = 'This is the error';
				var expectedResult = `Error: ${errorMessage}`;
				var Api = require('../index');
				var api = new Api();
				api.get('/test', (request) => {
					throw new Error(errorMessage);
				});

				api.handler({
					httpMethod: 'GET',
					resource: '/test'
				}, {}, (_, x) => x)
				.then(output => {
					assert.deepEqual(JSON.parse(output.body), expectedResult, `Output data does not match expected.`);
					assert.strictEqual(output.statusCode, 500, 'Error should be a 500 on a throw');
					done();
				})
				.catch(failure => done(failure));
			}
			catch(e) {
				console.error(e.stack);
				assert(false, e.toString());
			}
		});
	});
});

tests/make.test.js
const { describe, it } = require('mocha');

describe('make.js', () => {
  describe('Syntax', () => {
    it('Should be valid node', () => {
      require('../make');
    });
  });
});


make.js
/**
 * Module dependencies
 */
 let commander = require('commander');
 let fs = require('fs-extra');

 const githubActionsRunner = require('ci-build-tools')(process.env.GITHUB_TOKEN);
 function getVersion() {
   let release_version = '0.0';
   const pull_request = '';
   const branch = process.env.GITHUB_REF;
   const build_number = `${process.env.GITHUB_RUN_NUMBER}`;

   //Builds of pull requests
   if (pull_request && !pull_request.match(/false/i)) {
     release_version = `0.${pull_request}`;
   } else if (!branch || !branch.match(/^(refs\/heads\/)?release[/-]/i)) {
     //Builds of branches that aren't master or release
     release_version = '0.0';
   } else {
     //Builds of release branches (or locally or on server)
     release_version = branch.match(/^(?:refs\/heads\/)?release[/-](\d+(?:\.\d+){0,3})$/i)[1];
   }
   return `${release_version}.${(build_number || '0')}.0.0.0.0`.split('.').slice(0, 3).join('.');
 }
 const version = getVersion();
 commander.version(version);

 /**
  * Build
  */
 commander
 .command('setup')
 .description('Setup require build files for npm package.')
 .action(async () => {
   let package_metadata = require('./package.json');
   package_metadata.version = version;
   await fs.writeJson('./package.json', package_metadata, { spaces: 2 });

   console.log('Building package %s (%s)', package_metadata.name, version);
   console.log('');
 });

 /**
  * After Build
  */
 commander
 .command('after_build')
 .description('Publishes git tags and reports failures.')
 .action(() => {
   let package_metadata = require('./package.json');
   console.log('After build package %s (%s)', package_metadata.name, version);
   console.log('');
   githubActionsRunner.PublishGitTag(version);
   githubActionsRunner.MergeDownstream('release/', 'main');
 });

 commander.on('*', () => {
   if (commander.args.join(' ') === 'tests/**/*.js') { return; }
   console.log(`Unknown Command: ${commander.args.join(' ')}`);
   commander.help();
   process.exit(0);
 });
 commander.parse(process.argv[2] ? process.argv : process.argv.concat(['build']));


index.d.ts
export namespace OpenApi {
  export interface ApiOptions {
      requestMiddleware?: ((event: object, context: object) => Promise<unknown>) | (() => unknown) | undefined;
      responseMiddleware?: ((request: object, context: object) => Promise<unknown>) | (() => unknown) | undefined;
      errorMiddleware?: ((event: object, error: object) => Promise<unknown>) | (() => unknown) | undefined;
  }

  export interface HttpMethodOptions {
      rawBody?: boolean | undefined;
  }

  export interface HttpResponse {
      statusCode?: number | undefined;
      headers?: object | undefined;
      body?: (object | string) | undefined;
  }
}

export class OpenApi {
  constructor(options: OpenApi.ApiOptions, overrideLogger?: () => void);

  setAuthorizer(authorizerFunc: (req?: unknown) => Promise<unknown>): void;
  onEvent(onEventFunc: (req?: unknown) => Promise<unknown>): void;
  onSchedule(onScheduleFunc: (req?: unknown) => Promise<unknown>): void;

  /**
   * @returns The the Path map { [path]: { [method]: {metadata} } } for discovery usage.
   */
  getPathMap(): Record<string, Record<string, unknown>>;

  /**
   * @returns The matched route based on the method and the specified path.
   */
  resolveRoute(method: string, path: string): string;

  head(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  head(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  get(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  get(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  post(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  post(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  put(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  put(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  patch(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  patch(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  query(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  query(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  delete(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  delete(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  options(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  options(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  any(route: string, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;
  any(route: string, options: OpenApi.HttpMethodOptions, handler: (req?: object) => (OpenApi.HttpResponse | Promise<OpenApi.HttpResponse>)): void;

  handler(event: object, context: object): Promise<unknown>;
}


src/response.js
class Response {
  constructor(body, statusCode, headers) {
    let pullBody = body && (body.body || body.statusCode || body.headers);
    this.statusCode = (pullBody ? body.statusCode : statusCode) || 200;
    let populatedHeaders = (pullBody ? body.headers : headers) || {};
    this.isBase64Encoded = (pullBody ? body.isBase64Encoded : false) || false;
    this.body = (this.statusCode === 204 || this.statusCode === 304) ? undefined : (pullBody ? body.body : body);

    if (!this.body) {
      delete this.body;
      populatedHeaders = Object.assign({ 'Access-Control-Allow-Origin': '*' }, populatedHeaders);
    } else if (this.body && this.body instanceof Buffer) {
      populatedHeaders = Object.assign({ 'Content-Type': 'application/octet-stream', 'Access-Control-Allow-Origin': '*' }, populatedHeaders);
    } else {
      this.body = this.isBase64Encoded || typeof this.body !== 'object' ? this.body : JSON.stringify(this.body);
      populatedHeaders = Object.assign({ 'Content-Type': 'application/links+json', 'Access-Control-Allow-Origin': '*' }, populatedHeaders);
    }

    this.multiValueHeaders = Object.keys(populatedHeaders).reduce((agg, h) => {
      agg[h] = Array.isArray(populatedHeaders[h]) ? populatedHeaders[h] : [populatedHeaders[h]].filter(v => v);
      return agg;
    }, {});
  }
}

module.exports = Response;


index.js
/* eslint-disable @typescript-eslint/no-empty-function */
const Resp = require('./src/response');
const PathResolver = require('./src/pathResolver');
let isFunction = obj => { return !!(obj && obj.constructor && obj.call && obj.apply); };

// This class is a singleton because the standard usage dereferences the `handler` method which removes the binding to `this`
let apiFactory = null;

class ApiFactory {
  constructor(options, overrideLogger = null) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    apiFactory = this;

    this.Authorizer = null;
    this.requestMiddleware = options && options.requestMiddleware || (r => r);
    this.responseMiddleware = options && options.responseMiddleware || ((_, r) => r);
    this.errorMiddleware = options && options.errorMiddleware || ((_, e) => e);
    this.debug = options && !!options.debug;
    this.handlers = {
      onEvent() {},
      onSchedule() {}
    };
    this.Routes = {};
    this.ProxyRoutes = {};
    this.paths = {};
    this.pathResolver = options && options.pathResolver || new PathResolver();
    this.logger = overrideLogger || (message => console.log(JSON.stringify(message, null, 2)));
  }

  setAuthorizer(authorizerFunc) {
    if (!isFunction(authorizerFunc)) { throw new Error('Authorizer Function has not been defined as a function.'); }
    this.Authorizer = authorizerFunc;
  }

  SetAuthorizer(authorizerFunc) {
    this.setAuthorizer(authorizerFunc);
  }

  onEvent(onEventFunc) {
    if (!isFunction(onEventFunc)) { throw new Error('onEvent has not been defined as a function.'); }
    this.handlers.onEvent = onEventFunc;
  }

  onSchedule(onScheduleFunc) {
    if (!isFunction(onScheduleFunc)) { throw new Error('onSchedule has not been defined as a function.'); }
    this.handlers.onSchedule = onScheduleFunc;
  }

  head(route, p0, p1) { this.method('HEAD', route, p0, p1); }
  get(route, p0, p1) { this.method('GET', route, p0, p1); }
  post(route, p0, p1) { this.method('POST', route, p0, p1); }
  put(route, p0, p1) { this.method('PUT', route, p0, p1); }
  patch(route, p0, p1) { this.method('PATCH', route, p0, p1); }
  query(route, p0, p1) { this.method('QUERY', route, p0, p1); }
  delete(route, p0, p1) { this.method('DELETE', route, p0, p1); }
  options(route, p0, p1) { this.method('OPTIONS', route, p0, p1); }
  any(route, p0, p1) { this.method('ANY', route, p0, p1); }

  method(verb, route, p0, p1) {
    let params = [p0, p1].filter(p => p);
    let handler = null;
    let options = {};
    if (params.length > 0) { handler = params[params.length - 1]; }
    if (params.length > 1) { options = params[params.length - 2]; }
    if (params.length > 2) { throw new Error(`Method not defined with ${params.length} parameters.  Closest match is function(route, options, handler).`); }

    if (!isFunction(handler)) { throw new Error('Handler is not defined as a function.'); }

    let path = route.toString();
    if (path[0] !== '/') { path = `/${route}`; }

    let api = {
      ResourcePath: path,
      Method: verb,
      Handler: handler,
      Options: options || {}
    };
    if (!apiFactory.Routes[path]) {
      apiFactory.Routes[path] = {};
      apiFactory.ProxyRoutes[path] = {};
    }
    apiFactory.Routes[path][verb] = api;
    apiFactory.ProxyRoutes = apiFactory.pathResolver.storePath(apiFactory.ProxyRoutes, verb, path, api);
    if (!apiFactory.paths[path]) {
      apiFactory.paths[path] = {};
    }
    if (verb !== 'ANY') {
      apiFactory.paths[path][verb] = {};
    }
  }

  getPathMap() {
    return apiFactory.paths;
  }

  resolveRoute(method, path) {
    const routeData = apiFactory.pathResolver.resolvePath(apiFactory.ProxyRoutes, method, path);
    return routeData && routeData.constructedRoute;
  }

  convertEvent(event) {
    event.openApiOptions = event.openApiOptions || {};
    event.queryStringParameters = event.queryStringParameters || {};
    event.stageVariables = event.stageVariables || {};
    event.pathParameters = event.pathParameters || {};

    if (!event.requestContext) {
      event.requestContext = {};
    }
    event.requestContext.authorizer = event.requestContext.authorizer || {};

    const method = event.httpMethod || event.requestContext.http && event.requestContext.http.method;
    let definedRoute = null;

    const proxyPath = '/{proxy+}';
    event.path = event.path || event.requestContext.http && event.requestContext.http.path || event.requestContext.path;
    // Remove stage from Path
    event.path = event.path.startsWith(`/${event.requestContext.stage}`) ? event.path.substring(event.requestContext.stage.length + 1) : event.path;

    // The replace handles cases where the route key is prepended with a method from API Gateway
    const routeKey = event.routeKey && event.routeKey.replace(/^[A-Z]+\s/, '') || event.resource;

    const map = apiFactory.pathResolver.resolvePath(apiFactory.ProxyRoutes, method, event.path);
    const definedMethods = map && map.methods;

    // default to defined path when proxy is not specified.
    if (routeKey.lastIndexOf(proxyPath) === -1 && routeKey !== '$default') {
      if (apiFactory.Routes[routeKey] && apiFactory.Routes[routeKey][method]) {
        definedRoute = apiFactory.Routes[routeKey][method];
      } else if (apiFactory.Routes[routeKey] && apiFactory.Routes[routeKey].ANY) {
        definedRoute = apiFactory.Routes[routeKey].ANY;
      }
    } else if (map) {
      // if it is a proxy path then then look up the proxied value.
      definedRoute = map.value;
      delete event.pathParameters.proxy;
      event.pathParameters = Object.assign({}, map.tokens, event.pathParameters);
    }

    // either it is proxied and not defined or not defined, either way go to the proxy method.
    if (!definedRoute) {
      if (apiFactory.Routes[proxyPath] && apiFactory.Routes[proxyPath][method]) {
        definedRoute = apiFactory.Routes[proxyPath][method];
      } else if (apiFactory.Routes[proxyPath] && apiFactory.Routes[proxyPath].ANY) {
        definedRoute = apiFactory.Routes[proxyPath].ANY;
      }
    }

    if (definedRoute) {
      event.route = definedRoute.ResourcePath;
      event.openApiOptions.definedMethods = definedMethods;
      return { event, definedRoute };
    }

    event.route = null;
    event.openApiOptions.definedMethods = definedMethods;
    return { event, definedRoute };
  }

  /* This is the entry point from AWS Lambda. */
  async handler(originalEvent, context) {
    if (apiFactory.debug) {
      apiFactory.logger({ level: 'DEBUG', title: 'Original Event, before transformation', originalEvent });
    }

    if ((originalEvent.path || originalEvent.rawPath) && !originalEvent.type) {
      let { event, definedRoute } = apiFactory.convertEvent(originalEvent);
      if (!definedRoute) {
        return new Resp({
          title: 'No handler defined for method and resource.',
          details: { event, context }
        }, 500);
      }

      let lambda = definedRoute.Handler;
      event.openApiOptions = Object.assign({}, event.openApiOptions, definedRoute.Options || {});
      if (event.isBase64Encoded) {
        event.body = Buffer.from(event.body || '', 'base64').toString('utf8');
        event.isBase64Encoded = false;
      }
      if (!definedRoute.Options.rawBody) {
        // Convert a string body into a javascript object, if it is valid json and raw body is not set.
        try {
          event.body = JSON.parse(event.body);
        } catch (e) { /* */ }
      }

      try {
        let request = await apiFactory.requestMiddleware(event, context);
        let response = await lambda(request, context);
        let result = await apiFactory.responseMiddleware(request, response);
        if (!result) { return new Resp(null, 204); }
        if (!(result instanceof Resp)) { return new Resp(result, result && result.statusCode ? null : 200); }
        return result;
      } catch (e) {
        try {
          let error = await apiFactory.errorMiddleware(event, e);
          if (error instanceof Resp) { return error; }
          if (error instanceof Error) {
            apiFactory.logger({ level: 'ERROR', title: 'Exception thrown by invocation of the runtime lambda function, check the implementation.', api: definedRoute, error: e });
            return new Resp({ title: 'Unexpected error', errorId: event.requestContext.requestId }, 500);
          }
          return new Resp(error, error && error.statusCode ? null : 500);
        } catch (middleE) {
          apiFactory.logger({ level: 'ERROR', title: 'Exception thrown by invocation of the error middleware, check the implementation.', api: definedRoute, error: e, middleware: middleE });
          return new Resp({ title: 'Unexpected error', errorId: event.requestContext.requestId }, 500);
        }
      }
    }

    //If this is the authorizer lambda, then call the authorizer
    if (originalEvent.type === 'REQUEST' && (originalEvent.methodArn || originalEvent.routeArn)) {
      if (!apiFactory.Authorizer) {
        apiFactory.logger({ title: 'No authorizer function defined' });
        throw new Error('Authorizer Undefined');
      }

      const { event } = apiFactory.convertEvent(originalEvent);
      try {
        let policy = await apiFactory.Authorizer(event, context);
        if (!policy.principalId) {
          apiFactory.logger({ title: 'OpenAPI-Factory: PolicyResult Failure, missing required parameter in policy: principalId', level: 'WARN', details: policy });
        }
        if (apiFactory.debug) {
          apiFactory.logger({ title: 'OpenAPI-Factory: PolicyResult Success', level: 'INFO', details: policy });
        }
        return policy;
      } catch (error) {
        if (apiFactory.debug) {
          apiFactory.logger({ title: 'OpenAPI-Factory: PolicyResult Failure', level: 'WARN', error });
        }
        throw error;
      }
    }

    // this is a scheduled trigger
    if (originalEvent.source === 'aws.events') {
      // eslint-disable-next-line no-return-await
      return await apiFactory.handlers.onSchedule(originalEvent, context);
    }

    // Otherwise execute the onEvent handler
    // eslint-disable-next-line no-return-await
    return await apiFactory.handlers.onEvent(originalEvent, context);
  }
}

ApiFactory.getInstance = function getInstance() {
  return apiFactory;
};

module.exports = ApiFactory;
